### PR TITLE
Create FoundationInternals, an internal module to host shared files used by FoundationInternationalization and FoundationEssentials

### PR DESCRIPTION
**Pull request by itingliu**
- Move `LockedState` to FoundationInternals and publicize functions needed by other modules.
- Modules will access FoundationInternals types with `@_implementationOnly import FoundationInternals`.
- Note that tests with `@testable import FoundationEssentials/FoundationInternationalization` will need to import FoundationInternals explicitly if the test requires FoundationInternals types.

### CODE REVIEW

**Comment by itingliu** [comment]
@swift-ci please test

**Comment by itingliu** [comment]
@swift-ci test

File: Sources/FoundationEssentials/AttributedString/AttributedString+Guts.swift [hunk 1]
```diff
@@ -16,6 +16,8 @@
 import _RopeModule
 #endif
 
+package import FoundationInternals
```
Comment (jmschonfeld) [inline Sources/FoundationEssentials/AttributedString/AttributedString+Guts.swift:19]: I think for this and the rest of the `FoundationInternal` imports, we'll either want to move them into the `!FOUNDATION_FRAMEWORK` section (the `else` block above) or surround it in a `#if canImport(FoundationInternals)`

**Comment by iCharlesHu** [comment]
Also if you rebase the Linux tests should now pass.

File: Sources/FoundationEssentials/AttributedString/AttributedString+Guts.swift (hunk 1 above)
Comment (itingliu) [reply Sources/FoundationEssentials/AttributedString/AttributedString+Guts.swift:20]: Thanks. Updated

File: Sources/FoundationEssentials/JSON/JSONDecoder.swift [hunk 2]
```diff
@@ -17,6 +17,7 @@ import Glibc
 #endif
 
 @_implementationOnly import _CShims
```
[the remark this answers is not in the document]
Comment (itingliu) [reply Sources/FoundationEssentials/JSON/JSONDecoder.swift]: I prefer do this in a separate PR

**Review by jmschonfeld — APPROVED** [review]

**Comment by itingliu** [comment]
@swift-ci please test

**Review by glessard — APPROVED** [review]

**Comment by itingliu** [comment]
@swift-ci please test

**Comment by iCharlesHu** [comment]
```
/Users/ec2-user/jenkins/workspace/pr-swift-foundation-macos/branch-main/swift-foundation/Sources/FoundationEssentials/JSON/JSONEncoder.swift:23:1: error: expected expression
@_objcRuntimeName(_TtC10Foundation13__JSONEncoder)
^
[220/562] Compiling FoundationEssentials StringProtocol+Stub.swift
/Users/ec2-user/jenkins/workspace/pr-swift-foundation-macos/branch-main/swift-foundation/Sources/FoundationEssentials/JSON/JSONEncoder.swift:23:1: error: expected expression
@_objcRuntimeName(_TtC10Foundation13__JSONEncoder)
```
What's this Linux failure about? We haven't touched that right?

File: Sources/FoundationEssentials/JSON/JSONEncoder.swift [hunk 3]
```diff
@@ -22,6 +22,11 @@
 #if FOUNDATION_FRAMEWORK
 @_objcRuntimeName(_TtC10Foundation13__JSONEncoder)
 #endif
+
+#if canImport(FoundationInternals)
+package import FoundationInternals
+#endif
+
```
Comment (iCharlesHu) [inline Sources/FoundationEssentials/JSON/JSONEncoder.swift:29]: Ahhh... I believe you'll have to move this above the `#if FOUNDATION_FRAMEWORK` bit.

Comment (itingliu) [reply Sources/FoundationEssentials/JSON/JSONEncoder.swift:29]: I suspected this is what caused that error, but do you know why the order would matter?

**Comment by itingliu** [comment]
@swift-ci please test

**Comment by itingliu** [comment]
@swift-ci please test

File: Sources/FoundationEssentials/JSON/JSONEncoder.swift (hunk 3 above)
Comment (itingliu) [reply Sources/FoundationEssentials/JSON/JSONEncoder.swift:29]: oh wait I see it now... that `@_objcRuntimeName(_TtC10Foundation13__JSONEncoder)`  is an attribute for `class JSONEncoder`

**Comment by itingliu** [comment]
@swift-ci please test

File: Sources/FoundationEssentials/JSON/JSONEncoder.swift (hunk 3 above)
Comment (iCharlesHu) [reply Sources/FoundationEssentials/JSON/JSONEncoder.swift]: Yes exactly.

**Comment by itingliu** [comment]
@swift-ci please test

**Pull request merged**